File: solutions/typescript/0037.ts
function solveSudoku(board: string[][]): void {
  // 9x9 数独，使用回溯 + 位掩码（bitmask）进行求解：
  // - 用三个掩码数组 rowMasks/colMasks/boxMasks 记录每行/列/宫中已使用的数字
  // - 空位按“候选数最少”优先（启发式）排序，降低回溯分支
  // - 每次取最低位 1 作为候选数字，放置 -> 递归 -> 回溯
  const N = 9;
  const FULL_MASK = (1 << 9) - 1; // 9 位 1，对应数字 1..9 的可用位

  const rowMasks: number[] = new Array(N).fill(0); // 第 i 行已使用位
  const colMasks: number[] = new Array(N).fill(0); // 第 j 列已使用位
  const boxMasks: number[] = new Array(N).fill(0); // 第 k 宫已使用位（k = r/3*3 + c/3）
  const empties: Array<[number, number]> = []; // 记录所有空格位置

  const getBoxIndex = (r: number, c: number): number => ((Math.floor(r / 3) * 3) + Math.floor(c / 3));

  // 初始化：统计已填数字到掩码中，并收集空格
  for (let r = 0; r < N; r++) {
    for (let c = 0; c < N; c++) {
      const ch = board[r][c];
      if (ch === '.') {
        empties.push([r, c]);
        continue;
      }
      const digit = ch.charCodeAt(0) - '1'.charCodeAt(0); // 将 '1'..'9' 映射到 0..8
      const bit = 1 << digit;
      rowMasks[r] |= bit;
      colMasks[c] |= bit;
      boxMasks[getBoxIndex(r, c)] |= bit;
    }
  }

  // 启发式：空格按候选数（可用位数量）升序排序，减少搜索空间
  empties.sort((a, b) => {
    const [ra, ca] = a;
    const [rb, cb] = b;
    const maskA = rowMasks[ra] | colMasks[ca] | boxMasks[getBoxIndex(ra, ca)];
    const maskB = rowMasks[rb] | colMasks[cb] | boxMasks[getBoxIndex(rb, cb)];
    const availA = FULL_MASK & ~maskA; // 可用位：1 表示该数字可填
    const availB = FULL_MASK & ~maskB;
    const countA = countBits(availA);
    const countB = countBits(availB);
    return countA - countB;
  });

  function countBits(x: number): number {
    // Kernighan 算法：每次清除最低位 1，统计次数
    let cnt = 0;
    while (x !== 0) {
      x &= (x - 1);
      cnt++;
    }
    return cnt;
  }

  function dfs(idx: number): boolean {
    // 若所有空格填完，找到解
    if (idx === empties.length) return true;

    const [r, c] = empties[idx];
    const used = rowMasks[r] | colMasks[c] | boxMasks[getBoxIndex(r, c)]; // 当前格不可用数字集合
    let avail = FULL_MASK & ~used; // 可用数字集合（bit 为 1 表示可用）

    while (avail !== 0) {
      const bit = avail & -avail; // 取最低位 1（选择一个候选）
      const digit = Math.log2(bit) | 0; // 对应数字 0..8
      const ch = String.fromCharCode('1'.charCodeAt(0) + digit);

      // 放置该候选
      board[r][c] = ch;
      rowMasks[r] |= bit;
      colMasks[c] |= bit;
      const boxIdx = getBoxIndex(r, c);
      boxMasks[boxIdx] |= bit;

      // 继续填下一个空格
      if (dfs(idx + 1)) return true;

      // 回溯：撤销放置
      board[r][c] = '.';
      rowMasks[r] &= ~bit;
      colMasks[c] &= ~bit;
      boxMasks[boxIdx] &= ~bit;

      // 去掉该候选，尝试下一个候选
      avail &= (avail - 1);
    }

    // 所有候选均失败，回溯上一层
    return false;
  }

  dfs(0);
}

File: solutions/rust/0037.rs
impl Solution {
    pub fn solve_sudoku(board: &mut Vec<Vec<char>>) {
        // 9x9 数独：使用回溯 + 位掩码
        // - row_masks/col_masks/box_masks 记录每行/列/宫已使用数字（bit=1 表示已用）
        // - 收集空格坐标，按候选数最少排序以减少回溯
        // - 逐格尝试可用数字，递归求解；失败则回溯
        const N: usize = 9;
        const FULL_MASK: u16 = (1 << 9) - 1; // 0b1_1111_1111

        let mut row_masks: [u16; N] = [0; N];
        let mut col_masks: [u16; N] = [0; N];
        let mut box_masks: [u16; N] = [0; N];
        let mut empties: Vec<(usize, usize)> = Vec::new();

        let get_box_index = |r: usize, c: usize| -> usize { (r / 3) * 3 + (c / 3) };

        // 初始化掩码与空位列表
        for r in 0..N {
            for c in 0..N {
                let ch = board[r][c];
                if ch == '.' {
                    empties.push((r, c));
                    continue;
                }
                let digit = (ch as u8 - b'1') as u16; // 0..8
                let bit = 1u16 << digit;
                row_masks[r] |= bit;
                col_masks[c] |= bit;
                box_masks[get_box_index(r, c)] |= bit;
            }
        }

        // 计数 1 比特个数（Kernighan）
        fn count_bits(mut x: u16) -> u32 {
            let mut cnt = 0u32;
            while x != 0 {
                x &= x - 1;
                cnt += 1;
            }
            cnt
        }

        // 启发式：空位按候选数升序排序
        empties.sort_by(|&(ra, ca), &(rb, cb)| {
            let mask_a = row_masks[ra] | col_masks[ca] | box_masks[get_box_index(ra, ca)];
            let mask_b = row_masks[rb] | col_masks[cb] | box_masks[get_box_index(rb, cb)];
            let avail_a = FULL_MASK & !mask_a;
            let avail_b = FULL_MASK & !mask_b;
            count_bits(avail_a).cmp(&count_bits(avail_b))
        });

        // DFS 回溯
        fn dfs(
            idx: usize,
            empties: &[(usize, usize)],
            board: &mut [Vec<char>],
            row_masks: &mut [u16; N],
            col_masks: &mut [u16; N],
            box_masks: &mut [u16; N],
        ) -> bool {
            if idx == empties.len() { return true; }

            let (r, c) = empties[idx];
            let box_idx = (r / 3) * 3 + (c / 3);
            let used = row_masks[r] | col_masks[c] | box_masks[box_idx];
            let mut avail = FULL_MASK & !used; // 可用数字位集合

            while avail != 0 {
                // 取最低位 1
                let bit = avail & (!avail + 1); // 等价于 avail & -avail（在无符号下的两补运算）
                let digit = bit.trailing_zeros() as u8; // 0..8
                let ch = (b'1' + digit) as char;

                // 放置
                board[r][c] = ch;
                row_masks[r] |= bit;
                col_masks[c] |= bit;
                box_masks[box_idx] |= bit;

                if dfs(idx + 1, empties, board, row_masks, col_masks, box_masks) { return true; }

                // 回溯
                board[r][c] = '.';
                row_masks[r] &= !bit;
                col_masks[c] &= !bit;
                box_masks[box_idx] &= !bit;

                // 去掉该候选
                avail &= avail - 1;
            }

            false
        }

        dfs(0, &empties, board, &mut row_masks, &mut col_masks, &mut box_masks);
    }
}